relayer/src/config.rs

```diff
@@ -405,6 +408,42 @@ pub struct ChainConfig {
     pub address_type: AddressType,
 }
 
+mod proof_specs_serde {
+    use core::fmt;
+    use ibc::core::ics23_commitment::specs::ProofSpecs;
+    use serde::{de, ser, Deserializer, Serializer};
+
+    pub fn serialize<S: Serializer>(
+        proof_specs: &ProofSpecs,
+        serializer: S,
+    ) -> Result<S::Ok, S::Error> {
+        let json_str = serde_json::to_string(proof_specs).map_err(ser::Error::custom)?;
+        serializer.serialize_str(&json_str)
```

Comment:
So this is forced to be a JSON string embedded into any serialized format, human-readable or not?

When the JSON serializer is used, this is will end up as a string with escaped JSON inside. Forgive the newbie, but I'd rather like to know a good reason why the serialization can't be structurally composed. Is it because we can't lock down the serialization schema for `ProofSpecs` in the embedding format?

Reply:
> So this is forced to be a JSON string embedded into any serialized format, human-readable or not?

Right, but this (i.e. `serde(with=proof_specs_serde)`) is only used in the config which is serialized as TOML. 

> When the JSON serializer is used for an outer structure, this is will end up as a string with escaped JSON inside. 

Nice point, I modified the serializing code to serialize as a pretty-printed string, and the TOML serializer defaults to pretty strings (represented with `'''`) to serialize strings with newlines in them. Also tested it locally.

Reply:
The serde impls are agnostic to the serializer and they are part the public API, so there's no stopping anyone from serializing in any other format. I agree it wouldn't make sense for the config to be serialized in something non-readable by humans, though.

To be clear, this does not just delegate to the `ProofSpecs`' serde impls to get it emitted as native JSON in JSON (in case someone wants to have a JSON config), because the primary use case is TOML and it probably would look horrible, unless you have a way to tell the toml-rs serializer to switch to inline JSON at the level of that value, or is there some other reason?

Reply:
Sheesh, I forgot that TOML is actively trying not to become YAML, and as such does not support inline JSON.

Reply:
Sorry, but I'm not sure I understand the concern here. The rationale for choosing to (de)serialize using `serde(with=proof_specs_serde)` was that `ProofSpecs` encoded as TOML would look ugly (and take a lot of space). Although `ProofSpecs` is in the public API, the `proof_specs_serde` module is private and meant to be used only for (de)serializing `ProofSpecs` within the `Config`.
Note that `ProofSpecs` also implements (derives) `Serialize`/`Deserialize` itself for other use cases.

Reply:
Aahaa, I get it now! Not sure what would be the best way to solve this (I think a proper solution would require specialization). Found this related issue -> [Serializer-specific struct serialization
](https://github.com/serde-rs/serde/issues/968). Do you have any ideas on how we can solve this problem? 
One good thing is that all components of the config are public themselves, so users can compose their own config types. Also, AFAIK the relayer crate is only meant to be used by other packages in the workspace.

Reply:
I think it's not really solvable within serde because it's designed to be largely agnostic about the serialization format (`is_human_readable` being the only concession).

It may be not such a big problem if the crate is not meant to be widely used.

Otherwise, the public API could hide the serde impls by implementing them on private structs, and lock down TOML as the only possible serialization format, but this requires a lot of newtype wrappers, accessor methods and so on. I don't think it's worth it.